File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
function solution(numbers) {
    const primeSet = new Set();
    const numLength = numbers.length;
    const visited = new Array(numLength).fill(false);
    
    const dfs = (current) => {    
        if (isPrime(current)) primeSet.add(current);
        
        for (let i = 0; i < numLength; i++) {
            if (!visited[i]) {
                visited[i] = true;
                dfs(current * 10 + parseInt(numbers[i]));
                visited[i] = false;
            }
        }
    }
    
    for (let i = 0; i < numLength; i++) {
        visited[i] = true;
        dfs(parseInt(numbers[i]));
        visited[i] = false;
    }
    
    return primeSet.size;
}

function isPrime(num) {
    if (num <= 1) return false;
    for (let i = 2; i <= Math.sqrt(num); i++) {
        if (num % i === 0) return false;
    }
    
    return true;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    private Set<Integer> primeSet = new HashSet<>();
    private boolean[] visited;
    private String numbers;
    private int numLength;

    public int solution(String numbers) {
        this.numbers = numbers;
        this.numLength = numbers.length();
        this.visited = new boolean[numLength];
        
        for (int i = 0; i < numLength; i++) {
            visited[i] = true;
            dfs(Integer.parseInt(String.valueOf(numbers.charAt(i))));
            visited[i] = false;
        }
        
        return primeSet.size();
    }

    private void dfs(int current) {
        if (isPrime(current)) {
            primeSet.add(current);
        }

        for (int i = 0; i < numLength; i++) {
            if (!visited[i]) {
                visited[i] = true;
                dfs(current * 10 + Integer.parseInt(String.valueOf(numbers.charAt(i))));
                visited[i] = false;
            }
        }
    }

    private boolean isPrime(int num) {
        if (num <= 1) return false;
        for (int i = 2; i <= Math.sqrt(num); i++) {
            if (num % i == 0) return false;
        }
        return true;
    }
}
